### PR TITLE
Add support for Qt 6

### PR DESCRIPTION
**Pull request by data-man**
- added `CONTOUR_BUILD_WITH_QT6` option
- `CONTOUR_BLUR_PLATFORM_KWIN` option will disabled for Qt 6

<details>
<summary>$ ldd contour</summary>

```
linux-vdso.so.1 (0x00007ffef5b05000)
libQt6Core5Compat.so.6 => /usr/lib/libQt6Core5Compat.so.6 (0x00007f17660d5000)
libQt6Network.so.6 => /usr/lib/libQt6Network.so.6 (0x00007f1765ede000)
libQt6OpenGLWidgets.so.6 => /usr/lib/libQt6OpenGLWidgets.so.6 (0x00007f1765ed1000)
libQt6Widgets.so.6 => /usr/lib/libQt6Widgets.so.6 (0x00007f17657a6000)
libutil.so.1 => /usr/lib/libutil.so.1 (0x00007f17657a1000)
libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f176579a000)
libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x00007f1765749000)
libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x00007f1765675000)
libharfbuzz.so.0 => /usr/lib/libharfbuzz.so.0 (0x00007f176559f000)
libQt6OpenGL.so.6 => /usr/lib/libQt6OpenGL.so.6 (0x00007f17654fd000)
libQt6Gui.so.6 => /usr/lib/libQt6Gui.so.6 (0x00007f1764cae000)
libGLX.so.0 => /usr/lib/libGLX.so.0 (0x00007f1764c7b000)
libOpenGL.so.0 => /usr/lib/libOpenGL.so.0 (0x00007f1764c4d000)
libQt6Core.so.6 => /usr/lib/libQt6Core.so.6 (0x00007f17646bd000)
libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f176469c000)
libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f1764486000)
libm.so.6 => /usr/lib/libm.so.6 (0x00007f1764341000)
libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f1764326000)
libc.so.6 => /usr/lib/libc.so.6 (0x00007f1764157000)
libicuuc.so.69 => /usr/lib/libicuuc.so.69 (0x00007f1763f64000)
libzstd.so.1 => /usr/lib/libzstd.so.1 (0x00007f1763e55000)
libgssapi_krb5.so.2 => /usr/lib/libgssapi_krb5.so.2 (0x00007f1763e00000)
libbrotlidec.so.1 => /usr/lib/libbrotlidec.so.1 (0x00007f1763df2000)
libz.so.1 => /usr/lib/libz.so.1 (0x00007f1763dd1000)
libssl.so.1.1 => /usr/lib/libssl.so.1.1 (0x00007f1763d3e000)
libcrypto.so.1.1 => /usr/lib/libcrypto.so.1.1 (0x00007f1763a60000)
/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f1766194000)
libexpat.so.1 => /usr/lib/libexpat.so.1 (0x00007f1763a30000)
libbz2.so.1.0 => /usr/lib/libbz2.so.1.0 (0x00007f1763a1d000)
libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007f17639e6000)
libgraphite2.so.3 => /usr/lib/libgraphite2.so.3 (0x00007f17639bf000)
libglib-2.0.so.0 => /usr/lib/libglib-2.0.so.0 (0x00007f176388a000)
libEGL.so.1 => /usr/lib/libEGL.so.1 (0x00007f1763878000)
libX11.so.6 => /usr/lib/libX11.so.6 (0x00007f1763737000)
libQt6DBus.so.6 => /usr/lib/libQt6DBus.so.6 (0x00007f176366d000)
libxkbcommon.so.0 => /usr/lib/libxkbcommon.so.0 (0x00007f1763625000)
libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x00007f176356b000)
libicui18n.so.69 => /usr/lib/libicui18n.so.69 (0x00007f1763244000)
libdouble-conversion.so.3 => /usr/lib/libdouble-conversion.so.3 (0x00007f176322d000)
libb2.so.1 => /usr/lib/libb2.so.1 (0x00007f176320f000)
libpcre2-16.so.0 => /usr/lib/libpcre2-16.so.0 (0x00007f1763183000)
libicudata.so.69 => /usr/lib/libicudata.so.69 (0x00007f1761628000)
libkrb5.so.3 => /usr/lib/libkrb5.so.3 (0x00007f1761541000)
libk5crypto.so.3 => /usr/lib/libk5crypto.so.3 (0x00007f1761510000)
libcom_err.so.2 => /usr/lib/libcom_err.so.2 (0x00007f176150a000)
libkrb5support.so.0 => /usr/lib/libkrb5support.so.0 (0x00007f17614fb000)
libkeyutils.so.1 => /usr/lib/libkeyutils.so.1 (0x00007f17614f4000)
libresolv.so.2 => /usr/lib/libresolv.so.2 (0x00007f17614d8000)
libbrotlicommon.so.1 => /usr/lib/libbrotlicommon.so.1 (0x00007f17614b5000)
libpcre.so.1 => /usr/lib/libpcre.so.1 (0x00007f176143e000)
libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007f1761414000)
libdbus-1.so.3 => /usr/lib/libdbus-1.so.3 (0x00007f17613bf000)
libgomp.so.1 => /usr/lib/libgomp.so.1 (0x00007f1761379000)
libXau.so.6 => /usr/lib/libXau.so.6 (0x00007f1761374000)
libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007f176136c000)
libelogind.so.0 => /usr/lib/libelogind.so.0 (0x00007f17612d6000)
libcap.so.2 => /usr/lib/libcap.so.2 (0x00007f17612cb000)
```
</details>

There are no checks in CI yet. :)

### CODE REVIEW

File: src/contour/CMakeLists.txt [hunk 1]
```diff
@@ -110,7 +116,11 @@ if(Boost_FILESYSTEM_FOUND)
 endif()
 
 if(CONTOUR_FRONTEND_GUI)
-    target_link_libraries(contour contour_frontend_opengl terminal_renderer Qt5::Gui Qt5::Widgets Qt5::Network)
+    if(CONTOUR_BUILD_WITH_QT6)
+        target_link_libraries(contour contour_frontend_opengl terminal_renderer Qt6::Core Qt6::Network Qt6::OpenGL Qt6::OpenGLWidgets Qt6::Widgets)
```
Comment (christianparpart) [inline src/contour/CMakeLists.txt:120]: `contour_frontend_opengl terminal_renderer` should be moved out of this if-condition and only qt-related libraries should be added here (or into a variable that is later on fed to `target_link_libraries()`.

File: Changelog.md [hunk 2]
```diff
@@ -56,6 +56,7 @@
 - Adds new CLI command: `contour generate config output OUTPUT_FILE` to create a new default config.
 - Adds new CLI command: `contour generate integration shell SHELL output OUTPUT_FILE` to create the shell integreation file for the given shell (only zsh supported for now). Also adds a pre-generated shell integration file for Linux (and OS/X) to `/usr/share/contour/shell-integration.zsh`.
 - Unicode data updated to version 14.0 beta. See https://home.unicode.org/unicode-14-0-beta-review.
+- Adds support for Qt 6.
```
Comment (christianparpart) [inline Changelog.md:59]:
```suggestion
- Adds support for building with Qt 6 (disabled by default).
```

File: src/contour/TerminalWindow.h [hunk 3]
```diff
@@ -25,7 +25,7 @@
 #include <QtCore/QPoint>
 #include <QtCore/QTimer>
 #include <QtGui/QOpenGLExtraFunctions>
-#include <QtGui/QOpenGLWindow>
+#include <QOpenGLWindow>
```
Comment (christianparpart) [inline src/contour/TerminalWindow.h:28]: Don't they namespace per directory anymore? If the rest is still working via subdir, maybe we can prefix the include here by qt module, too?

File: src/contour/opengl/OpenGLRenderer.h [hunk 4]
```diff
@@ -18,9 +18,9 @@
 
 #include <crispy/debuglog.h>
 
-#include <QtGui/QMatrix4x4>
-#include <QtGui/QOpenGLExtraFunctions>
-#include <QtGui/QOpenGLShaderProgram>
+#include <QMatrix4x4>
```
Comment (christianparpart) [inline src/contour/opengl/OpenGLRenderer.h:21]: Why doesn't this work anymore? (basically the same question as above :) )

**Review by christianparpart — CHANGES_REQUESTED** [review]
I'm surprised. No code changes! That's actually very good.

I'd highly suggest to also add a CI workflow task for building against Qt 6.

And maybe we can reestablish the module names in the include directives, so one knows where these include files are coming from (what qt module).

File: src/contour/CMakeLists.txt [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
-
 option(CONTOUR_PERF_STATS "Enables debug printing some performance stats." OFF)
 option(CONTOUR_VT_METRICS "Enables collecting and exit-printing some VT usage metrics." OFF)
 option(CONTOUR_SCROLLBAR "Enables scrollbar in GUI frontend." ON)
 option(CONTOUR_BLUR_PLATFORM_KWIN "Enables support for blurring transparent background when using KWin (KDE window manager)." OFF)
+option(CONTOUR_BUILD_WITH_QT6 "Use Qt 6" OFF)
+
+if(CONTOUR_BUILD_WITH_QT6)
+   find_package(Qt6 COMPONENTS Core Gui Network OpenGL OpenGLWidgets Widgets REQUIRED)
+   set(CONTOUR_BLUR_PLATFORM_KWIN OFF)
```
Comment (christianparpart) [inline src/contour/CMakeLists.txt:11]: Ah, and why is this? What's the reason behind disabling?

File: src/contour/TerminalWindow.h (hunk 3 above)
Comment (data-man) [reply src/contour/TerminalWindow.h:28]: > Don't they namespace per directory anymore?

Yes.
`QOpenGLShader`, `QOpenGLShaderProgram` and `QOpenGLWindow` now in `/usr/include/qt6/QtOpenGL`.

File: src/contour/opengl/OpenGLRenderer.h (hunk 4 above)
Comment (data-man) [reply src/contour/opengl/OpenGLRenderer.h:21]: Oh!
Don't now why I changed it. :(

File: src/contour/CMakeLists.txt (hunk 5 above)
Comment (data-man) [reply src/contour/CMakeLists.txt:11]: Can't compile with KF5. Maybe we should wait for KF6. :)

File: src/contour/TerminalWindow.h (hunk 3 above)
Comment (data-man) [reply src/contour/TerminalWindow.h:28]: Interestingly, prefixes are rarely used in `/usr/share/doc/qt6/examples`. :)

**Comment by data-man** [comment]
> I'd highly suggest to also add a CI workflow task for building against Qt 6.

I'll try tomorrow. :sleeping_bed:

File: src/contour/TerminalWindow.h (hunk 3 above)
Comment (christianparpart) [reply src/contour/TerminalWindow.h:28]: I do prefer them though in order to know what modules they are coming from. Qt examples don't use them most likely to express simplicity, whereas I prefer descriptability.

File: src/contour/CMakeLists.txt [hunk 6]
```diff
@@ -110,7 +116,13 @@ if(Boost_FILESYSTEM_FOUND)
 endif()
 
 if(CONTOUR_FRONTEND_GUI)
-    target_link_libraries(contour contour_frontend_opengl terminal_renderer Qt5::Gui Qt5::Widgets Qt5::Network)
+    if(CONTOUR_BUILD_WITH_QT6)
+        target_link_libraries(contour contour_frontend_opengl terminal_renderer Qt6::Core Qt6::Network Qt6::OpenGL Qt6::OpenGLWidgets Qt6::Widgets)
+        target_compile_definitions(contour PRIVATE CONTOUR_BUILD_WITH_QT6)
+        target_compile_definitions(contour_frontend_opengl PRIVATE CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/CMakeLists.txt:122]: Exposing this preproc define is not needed, see below. :-)

File: src/contour/TerminalWindow.h [hunk 7]
```diff
@@ -25,7 +25,11 @@
 #include <QtCore/QPoint>
 #include <QtCore/QTimer>
 #include <QtGui/QOpenGLExtraFunctions>
-#include <QtGui/QOpenGLWindow>
+#if defined(CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/TerminalWindow.h:28]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

File: src/contour/opengl/CMakeLists.txt [hunk 8]
```diff
@@ -3,7 +3,12 @@ if("${OpenGL_GL_PREFERENCE}" STREQUAL "")
     set(OpenGL_GL_PREFERENCE "GLVND")
 endif()
 
-find_package(Qt5 COMPONENTS Gui Widgets REQUIRED)  # apt install qtbase5-dev libqt5gui5
+if(CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/opengl/CMakeLists.txt:6]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

File: src/contour/opengl/OpenGLRenderer.h [hunk 9]
```diff
@@ -20,7 +20,11 @@
 
 #include <QtGui/QMatrix4x4>
 #include <QtGui/QOpenGLExtraFunctions>
-#include <QtGui/QOpenGLShaderProgram>
+#if defined(CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/opengl/OpenGLRenderer.h:23]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

File: src/contour/opengl/ShaderConfig.h [hunk 10]
```diff
@@ -16,8 +16,11 @@
 #include <crispy/debuglog.h>
 
 #include <fmt/format.h>
-
-#include <QOpenGLShaderProgram>
+#if defined(CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/opengl/ShaderConfig.h:19]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

File: src/contour/opengl/TerminalWidget.h [hunk 11]
```diff
@@ -28,7 +28,11 @@
 #include <QtCore/QTimer>
 #include <QtGui/QOpenGLExtraFunctions>
 #include <QtGui/QVector4D>
-#include <QtWidgets/QOpenGLWidget>
+#if defined(CONTOUR_BUILD_WITH_QT6)
```
Comment (christianparpart) [inline src/contour/opengl/TerminalWidget.h:31]:
```suggestion
#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
```

**Review by christianparpart — CHANGES_REQUESTED** [review]
Sorry for having thought about this after the first review. That's however how we're already testing for Qt version requirements in [contour/opengl/TerminalWidget.cpp](https://github.com/contour-terminal/contour/blob/master/src/contour/opengl/TerminalWidget.cpp#L235-L245).

File: src/contour/TerminalWindow.h (hunk 7 above)
Comment (data-man) [reply src/contour/TerminalWindow.h:28]: At first I tried this, but it requires including `<QtCore/QtGlobal>` to `ShaderConfig.h`. :)
Is it acceptable?

**Comment by data-man** [comment]
@christianparpart let's improve CI in other PR(s), after switching to CPM. What do you think?

File: src/contour/TerminalWindow.h (hunk 7 above)
Comment (christianparpart) [reply src/contour/TerminalWindow.h]: Yes. Damn I saw that too late.

**Pull request merged**